drivers/can/can_numaker.c

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright (c) 2023 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_numaker_canfd
+
+#include <zephyr/drivers/reset.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/can.h>
+#include <zephyr/drivers/can/can_mcan.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/clock_control_numaker.h>
+#include <zephyr/logging/log.h>
+#include <soc.h>
+#include <NuMicro.h>
+
+LOG_MODULE_REGISTER(can_numaker, CONFIG_CAN_LOG_LEVEL);
+
+/* Implementation notes
+ * 1. Use Bosch M_CAN driver (m_can) as backend
+ * 2. Need to modify can_numaker_get_core_clock() for new SOC support
+ */
+
+struct can_numaker_config {
+	mm_reg_t   canfd_base;
+	mem_addr_t  mrba;
+	mem_addr_t  mram;
+	const struct reset_dt_spec reset;
+	uint32_t    clk_modidx;
+	uint32_t    clk_src;
+	uint32_t    clk_div;
+	const struct device *clk_dev;
+	void (*irq_config_func)(const struct device *dev);
+	const struct pinctrl_dev_config *pincfg;
+};
+
+static int can_numaker_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	const struct can_mcan_config *mcan_config = dev->config;
+	const struct can_numaker_config *config = mcan_config->custom;
+	uint32_t clksrc_rate_idx;
+	uint32_t clkdiv_divider;
+
+	/* Module clock source rate */
+	clksrc_rate_idx = CLK_GetModuleClockSource(config->clk_modidx);
+	/* Module clock divider */
+	clkdiv_divider = CLK_GetModuleClockDivider(config->clk_modidx) + 1;
+
+	switch (clksrc_rate_idx) {
+	case 0:        /* HXT */
+		*rate = __HXT / clkdiv_divider;
+		break;
+	case 1:        /* PLLFOUT/2 */
+		*rate = (CLK_GetPLLClockFreq() / 2) / clkdiv_divider;
+		break;
+	case 2:        /* HCLK */
+		*rate = CLK_GetHCLKFreq() / clkdiv_divider;
+		break;
+	case 3:        /* HIRC */
+		*rate = __HIRC / clkdiv_divider;
+		break;
+	default:
+		LOG_ERR("Invalid clock source rate index");
+		return -EIO;
+	}
+
+	LOG_DBG("Clock rate index/divider: %d/%d", clksrc_rate_idx, clkdiv_divider);
+
+	return 0;
+}
+
+static int can_numaker_init(const struct device *dev)
```

Comment:
Just a hint:
You can put the section that needs the registers unlocked in a static inline function and call that from the init function. This way you can just return from the function in any error case without the goto.
It is also much safer, because there is no way to exit without locking.

```C
static inline int init_locked(...) {
    ret = do_something()
    if(ret) {
        return ret;
    }

    ret = do_something()
    if(ret) {
        return ret;
    }

}

int init(...)
{
    unlock_registers();
    ret = init_locked();
    lock_registers();
    return ret(
}
```